### PR TITLE
Render ECE buttons

### LINKED ISSUE #8772: Render the payment buttons using ECE behind a feature flag.
Implement the rendering of payment buttons using the Express Checkout Element (ECE), controlled by a new feature flag. This involves setting up the feature flag and ensuring the ECE buttons render correctly when the flag is enabled.

### Expected behavior:

- With the feature flag enabled, payment buttons render using ECE without issues.
- With the feature flag disabled, payment buttons use the existing rendering method (PRBs).
- It's not expected to be able to process a payment using the ECE implementation.

### Tasks 

- [x] Create the feature flag
- [ ] Implement ECE Payment Buttons using the [Stripe ECE documentation](https://docs.stripe.com/elements/express-checkout-element). Adjust the button rendering logic to conditionally use ECE based on the feature flag.

### LINKED ISSUE #8825: Refactor codebase to decouple PRB utils dependencies
We need to refactor the codebase to remove dependencies on `payment-request/utils` within the WooPay buttons and all other areas outside of Payment Request Buttons (PRBs). This change will allow us to enable or disable PRBs without affecting the stability of other parts of the application.

### PR DESCRIPTION
**Pull request by rafaelzaleski**
Closes #8772 
Closes #8825 

#### Changes proposed in this Pull Request

Render the ECE buttons behind a feature flag. The buttons are rendered in:

- The product page.
- The shortcode cart/checkout.
- The blocks cart/checkout.

It also relocates functions that were used by other parts of the application, such as the WooPay button, from `payment-request/utils` to a new high-level utility directory, `utils/express-checkout`. This move aims to decouple these helpers from any specific express checkout implementation.

#### Testing instructions

Before you enable the feature flag, [test the PRB buttons](https://github.com/Automattic/woocommerce-payments/wiki/Testing-instructions-for-critical-flows#checkout-with-payment-request) and ensure there are no regressions. Also [test the WooPay express checkout button](https://github.com/Automattic/woocommerce-payments/wiki/Testing-instructions-for-critical-flows#woopay-checkout).

- Enable the ECE feature flag. `npm run wp option update _wcpay_feature_stripe_ece 1`
- Confirm that the ECE buttons are rendering on the shortcode and blocks product/cart/checkout pages.
- Ensure that the PRB buttons are not rendered when the feature flag is enabled.
- Ensure the ECE buttons are not rendered when the feature flag is disabled.

-------------------

- [x] Run `npm run changelog` to add a changelog file, choose `patch` to leave it empty if the change is not significant. You can add multiple changelog files in one PR by running this command a few times. 
- [ ] Covered with tests (or have a good reason not to test in description ☝️)
- [ ] Tested on mobile (or does not apply)

**Post merge**

- [ ] Link to testing instructions from [release testing doc](https://github.com/Automattic/woocommerce-payments/wiki/Release-testing-instructions) following [these instructions](https://github.com/Automattic/woocommerce-payments/wiki/How-to-write-good-manual-testing-scenarios) : _Add link here / 'QA Testing Not Applicable'_
- [ ] Add or update [critical flows](https://github.com/Automattic/woocommerce-payments/wiki/Critical-flows) and [testing instructions for critical flows](https://github.com/Automattic/woocommerce-payments/wiki/Testing-instructions-for-critical-flows), if applicable.
- [ ] Add what's changed (description, screenshot, demo videos etc.) to the release announcement post, if applicable.

### CODE REVIEW

**Comment by botwoo** [comment]
### Test the build 

#### Option 1. Jetpack Beta

- Install and activate [Jetpack Beta](https://jetpack.com/download-jetpack-beta/).
- Use this build by searching for PR number `8818` or branch name `feat/8772-render-ece-buttons` in your-test.site/wp-admin/admin.php?page=jetpack-beta&plugin=woocommerce-payments          

#### Option 2. Jurassic Ninja - available for logged-in A12s

:rocket: [Launch a JN site with this branch](https://jurassic.ninja/create/?jetpack-beta&shortlived&nojetpack&woocommerce&woocommerce-payments-dev-tools&branches.woocommerce-payments=feat/8772-render-ece-buttons) :rocket:

:information_source: Install this [Tampermonkey script](https://github.com/Automattic/woocommerce-payments/tree/develop/bin/wcpay-live-branches) to get more options.

---

Build info: 

- Latest commit: cb045a8c466e38c0d306428b75bdbdafc3ebe3bc
- Build time: 2024-05-17 20:10:56 UTC

_Note: the build is updated when a new commit is pushed to this PR._

File: client/express-checkout/blocks/index.js [hunk 1]
```diff
@@ -0,0 +1,23 @@
...
+
+/**
+ * Internal dependencies
+ */
+import { PAYMENT_METHOD_NAME_EXPRESS_CHECKOUT_ELEMENT } from '../../checkout/constants';
+import { ExpressCheckout } from './express-checkout';
+import { applePayImage } from './apple-pay-preview';
+import { getConfig } from '../../utils/checkout';
+
+const ApplePayPreview = () => <img src={ applePayImage } alt="" />;
```
Comment (cesarcosta99) [inline client/express-checkout/blocks/index.js]: Perhaps would be better to declare and export the `ApplePayPreview` component inside the `apple-pay-preview` file?

File: includes/express-checkout/class-wc-payments-express-checkout-button-handler.php [hunk 2]
```diff
@@ -0,0 +1,555 @@
...
+ */
+
+use WCPay\Exceptions\Invalid_Price_Exception;
+
+if ( ! defined( 'ABSPATH' ) ) {
+	exit;
+}
+
+use WCPay\Fraud_Prevention\Fraud_Prevention_Service;
+use WCPay\Logger;
```
Comment (cesarcosta99) [inline includes/express-checkout/class-wc-payments-express-checkout-button-handler.php]: Suggestion to put all imports together.

```suggestion
if ( ! defined( 'ABSPATH' ) ) {
	exit;
}

use WCPay\Exceptions\Invalid_Price_Exception;
use WCPay\Fraud_Prevention\Fraud_Prevention_Service;
use WCPay\Logger;
```

File: client/express-checkout/index.js [hunk 3]
```diff
@@ -0,0 +1,628 @@
...
+			return new Promise( ( resolve, reject ) => {
+				jQuery.post( url, args ).then( resolve ).fail( reject );
+			} );
+		}
+	);
+
+	let paymentRequestType;
+
+	// Track the payment request button click event.
+	const trackPaymentRequestButtonClick = ( source ) => {
```
Comment (cesarcosta99) [inline client/express-checkout/index.js]: Should we rename this to `trackExpressCheckoutButtonClick`?

File: client/express-checkout/index.js [hunk 4]
```diff
@@ -0,0 +1,628 @@
...
+		},
+		// A promise-based interface to jQuery.post.
+		( url, args ) => {
+			return new Promise( ( resolve, reject ) => {
+				jQuery.post( url, args ).then( resolve ).fail( reject );
+			} );
+		}
+	);
+
+	let paymentRequestType;
```
Comment (cesarcosta99) [inline client/express-checkout/index.js]: Should we rename this to `expressCheckoutType`?

File: client/express-checkout/index.js [hunk 5]
```diff
@@ -0,0 +1,628 @@
...
+
+	// Track the payment request button click event.
+	const trackPaymentRequestButtonClick = ( source ) => {
+		const paymentRequestTypeEvents = {
```
Comment (cesarcosta99) [inline client/express-checkout/index.js]: Should we rename this to `expressCheckoutTypeEvents`?

File: client/express-checkout/index.js [hunk 6]
```diff
@@ -0,0 +1,628 @@
...
+
+		/**
+		 * Creates Stripe Express Checkout Element.
+		 *
+		 * @param {Object} elements       Stripe elements instance.
+		 * @param {Object} options 		  Options for creating the Express Checkout Element.
+		 *
+		 * @return {Object} Stripe Express Checkout Element.
+		 */
+		createPaymentRequestButton: ( elements, options ) => {
```
Comment (cesarcosta99) [inline client/express-checkout/index.js]: Should we rename this to `createExpressCheckoutButton`

File: client/express-checkout/index.js [hunk 7]
```diff
@@ -0,0 +1,628 @@
...
+		if ( paymentRequestTypeEvents.hasOwnProperty( paymentRequestType ) ) {
+			const event = paymentRequestTypeEvents[ paymentRequestType ];
+			recordUserEvent( event, { source } );
+		}
+	};
+
+	/**
+	 * Object to handle Stripe payment forms.
+	 */
+	const wcpayPaymentRequest = {
```
Comment (cesarcosta99) [inline client/express-checkout/index.js]: Should we rename this to `wcpayExpressCheckout`?

File: client/express-checkout/index.js [hunk 8]
```diff
@@ -0,0 +1,628 @@
...
+
+			return api.paymentRequestAddToCart( data );
+		},
+
+		/**
+		 * Starts the payment request
+		 *
+		 * @param {Object} options Payment request options.
+		 */
+		startPaymentRequest: ( options ) => {
```
Comment (cesarcosta99) [inline client/express-checkout/index.js]: Should we rename this to `startExpressCheckout`?

File: client/express-checkout/index.js [hunk 9]
```diff
@@ -0,0 +1,628 @@
...
+		 * @param {Object} elements       Stripe elements instance.
+		 * @param {Object} options 		  Options for creating the Express Checkout Element.
+		 *
+		 * @return {Object} Stripe Express Checkout Element.
+		 */
+		createPaymentRequestButton: ( elements, options ) => {
+			return elements.create( 'expressCheckout', options );
+		},
+
+		attachPaymentRequestButtonEventListeners: (
```
Comment (cesarcosta99) [inline client/express-checkout/index.js]: I don't want to sound too repetitive here but from this point forward, we still have a lot of references to "payment request" that we may want to rename.

**Review by cesarcosta99 — CHANGES_REQUESTED** [review]
Code looks good overall! Before I dive into testing, I wanted to point out some suggestions. Mostly are about renaming some methods and variables, which is reasonable since we're splitting the approaches.

File: client/express-checkout/index.js (hunk 3 above)
Comment (rafaelzaleski) [reply client/express-checkout/index.js]: It's unclear how much of this code will be retained once we start processing payments. 
I also plan to refactor these files once I jump into payment processing. I'll assess whether it's easier to delete everything that is currently unused or simply rename everything 🤔

Comment (rafaelzaleski) [reply client/express-checkout/index.js]: I've updated some variable names and cleaned up the JavaScript to achieve a near-minimal implementation. I've requested another review.

File: client/express-checkout/blocks/apple-pay-preview.js [hunk 10]
```diff
@@ -0,0 +1,4 @@
...
+export const ApplePayPreview = () => <img src={ applePayImage } alt="" />;
```
Comment (cesarcosta99) [inline client/express-checkout/blocks/apple-pay-preview.js]: Here I'd export the component as default as this is the main and only component of this file.

```suggestion

const ApplePayPreview = () => <img src={ applePayImage } alt="" />;

export default ApplePayPreview;
```

File: client/express-checkout/blocks/index.js [hunk 11]
```diff
@@ -0,0 +1,22 @@
+/* global wcpayConfig */
+
+/**
+ * Internal dependencies
+ */
+import { PAYMENT_METHOD_NAME_EXPRESS_CHECKOUT_ELEMENT } from '../../checkout/constants';
+import { ExpressCheckout } from './express-checkout';
+import { ApplePayPreview } from './apple-pay-preview';
```
Comment (cesarcosta99) [inline client/express-checkout/blocks/index.js]: By updating how that component is exported, we need to update the import too.

```suggestion
import ApplePayPreview from './apple-pay-preview';
```

**Review by cesarcosta99 — CHANGES_REQUESTED** [review]
Thanks for the latest changes and cleaning things up a bit. I've tested it and everything worked as expected, except for ApplePay which I can't test it. I left one  last (and minor 🤏) suggestion before giving the final approval.

**Before enabling ECE**

- Checkout with GooglePay at the checkout page ✅ 
- Checkout with GooglePay in the product page ✅
- Checkout with ApplePay ⚠️
    - I don't have ApplePay set up, if needed, feel free to ask someone else from the team to test it.
- Checkout with the WooPay button from the checkout page ✅
- Checkout with the WooPay button from the product page ✅
- Ensure the ECE buttons are not rendered when the feature flag is disabled ✅
    - Checked product page, shortcode cart, shortcode checkout, cart Block and checkout Block.

**After enabling ECE**

- Confirm that the ECE buttons are rendering ✅
    - Checked product page, shortcode cart, shortcode checkout, cart Block and checkout Block.
- Ensure that the PRB buttons are not rendered ✅
    - Checked product page, shortcode cart, shortcode checkout, cart Block and checkout Block.

**Review by cesarcosta99 — APPROVED** [review]
LGTM :shipit:

**Pull request merged**